design-proposals/klip-24-key-column-semantics-in-queries.md

```diff
@@ -0,0 +1,372 @@
+
+## Motivation and background
+
+Below, we contrast the differences between transient and persistent queries and their current key
+column handling semantics.
+
+Note: in all the example sql in this KLIP:
+ * `EMIT CHANGES` had been removed for brevity.
+ * `ROWTIME` had been ignored for brevity.
+ * Schemas use `KEY` to denote the column stored in the Kafka record's key.
+ * `PRIMARY KEY` is not used for brevity.
+ * All input sources use the schema `ID INT KEY, V0 INT, V1 INT`, unless otherwise stated.
+
+Transient push and pull queries only return the columns in their projection. This is inline with
+standard SQL. Conversely, persistent queries _implicitly_ copy the key column across, and don't
+allow the key column in the projection.
+
+### Simple query
+
+The difference in key column handling can be seen with a simple query:
+
+```sql
+SELECT ID, V0, V1 FROM INPUT;
+-- resulting columns: ID INT, V0 INT, V1 INT
+
+-- vs --
+
+CREATE TABLE OUTPUT AS SELECT ID, V0, V1 FROM INPUT;
+-- fails with error about duplicate 'ID' column.
+```
+
+Where as the above transient query works and returns all columns, converting it to a persistent
+query causes it to fail with a duplicate column error, because `ID` is _implicitly_ being copied to
+the OUTPUT schema, and the `ID` in the projection is creating a value column called `ID`. That's two
+columns named `ID` and hence the query fails.
+
+We propose that the above `CREATE TABLE` statement should work, as the projection is equivalent to
+`select *`, which works.
+
+### Select star
+
+The differences with a 'select *' projection are more subtle.
+
+For non-join queries, both transient and persistent queries select all the columns in the schema:
+
+```sql
+SELECT * FROM INPUT;
+-- resulting columns: ID INT, V0 INT, V1 INT
+
+-- vs --
+
+CREATE TABLE OUTPUT AS SELECT * FROM INPUT;
+-- resulting schema: ID INT KEY, V0 INT, V1 INT (Same as above)
+```
+
+With a join query, a transient query selects all columns from all sources. A persistent query adds
+all columns from all sources, but also adds an additional copy of the left key column.
+
+```sql
+SELECT * FROM INPUT I1 JOIN INPUT_2 I2 ON I1.ID = I2.ID;
+-- resulting columns: I1_ID INT, I1_V0 INT, I1_V1 INT, I2_ID INT, I2_V0 INT, I2_V1 INT
+
+-- vs --
+
+CREATE TABLE OUTPUT AS SELECT * FROM INPUT I1 JOIN INPUT_2 I2 ON I1.ID = I2.ID;
+-- resulting schema: ID INT KEY, I1_ID INT, I1_V0 INT, I1_V1 INT, I2_ID INT, I2_V0 INT, I2_V1 INT
+```
+
+Note the addition of an additional `ID` column in the case of the persistent query.
+
+We propose that the join should not duplicate the left join column `I1.ID` into both the `ID` key
+and `I1_ID` value column.
+
+Only the key column should exist and it should be called either (TBD):
+  - `ID`, or
+  - `I1_ID`.
+
+Additionally, we may want to consider dropping `I2.ID` from the result, given we know it is equal
+to `I1.ID`. This effectively de-duplicates the data and saves on resources. Whether we do this is
+TBD.
```

Comment:
I don't think we should do this, it seems like other sql systems keeps the duplicate columns:
```sql
postgres=# CREATE TABLE input (id VARCHAR, count INTEGER);
postgres=# CREATE TABLE in2 (id VARCHAR, cost INTEGER);;
postgres=# SELECT * FROM input JOIN in2 ON input.id = in2.id;
 id | count | id | cost
----+-------+----+------
 a  |     1 | a  |  100
```

Reply:
I guess the reason we know the two are equal is that we only support `=` on the join criteria., where as other sql systems are more flexible.

If we remove the other column, then we may be painting ourselves into a corner, if in the future we can support more than `=`.

Good catch - I'll remove!